File: java/poker/src/main/java/com/duchen/poker/model/type/ThreeTakeTwoLineTypeChecker.java
package com.duchen.poker.model.type;

import com.duchen.poker.model.TwoValue;

import java.util.Arrays;
import java.util.List;

public class ThreeTakeTwoLineTypeChecker implements TypeChecker {

    @Override
    public CardGroupType getType() {
        return CardGroupType.THREE_TAKE_TWO_LINE;
    }

    @Override
    public List<Integer> getMyPossibleCardCount() {
        return Arrays.asList(10, 15, 20);
    }

    @Override
    public int getMaxCardIfIsMyType(List<Integer> cards) {
        int[] cardCountArray = new int[18];
        for (int card : cards) {
            cardCountArray[card]++;
        }

        int lineCount = 0;
        int doubleCount = 0;
        int i = 3, j = 3;
        for (; i < 15; i++) {
            if (cardCountArray[i] == 3) {
                lineCount++;
            } else {
                if (lineCount != 0) {
                    break;
                }
            }
        }
        for (; j < 16; j++) {
            if (cardCountArray[j] == 2 || cardCountArray[j] == 4) {
                doubleCount += cardCountArray[i] / 2;
            }
        }
        if (lineCount == doubleCount && lineCount * 5 == cards.size()) {
            return i - 1;
        } else {
            return -1;
        }
    }
}

File: java/poker/src/main/kotlin/poker/model/type/ThreeTakeTwoLineTypeChecker.kt
package com.duchen.poker.model.type

import com.duchen.poker.model.TwoValue

import java.util.Arrays

class ThreeTakeTwoLineTypeChecker : TypeChecker {

    override val type: CardGroupType
        get() = CardGroupType.THREE_TAKE_TWO_LINE

    override val myPossibleCardCount: List<Int>
        get() = Arrays.asList(10, 15, 20)

    override fun getMaxCardIfIsMyType(cards: MutableList<Int>): Int {
        val cardCountArray = IntArray(18)
        for (card in cards) {
            cardCountArray[card]++
        }

        var lineCount = 0
        var doubleCount = 0
        var i = 3
        var j = 3
        while (i < 15) {
            if (cardCountArray[i] == 3) {
                lineCount++
            } else {
                if (lineCount != 0) {
                    break
                }
            }
            i++
        }
        while (j < 16) {
            if (cardCountArray[j] == 2 || cardCountArray[j] == 4) {
                doubleCount += cardCountArray[i] / 2
            }
            j++
        }
        return if (lineCount == doubleCount && lineCount * 5 == cards.size) {
            i - 1
        } else {
            -1
        }
    }
}
